### PR TITLE
Method to identify injured segments to repair

### PR DESCRIPTION
**Pull request by jenlij**

### CODE REVIEW

**Comment by aleitner** [comment]
Can you update the checker to initialize and run inside the checker.Config.Run()
Same for the Queue.

File: pkg/datarepair/checker/checker.go [hunk 1]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+
+
+// Checker contains the information needed to do checks for missing pieces
+type Checker struct {
+	params *pb.IdentifyRequest
+	pointerdb storage.KeyValueStore
+	repairQueue *queue.Queue
+	overlay pb.OverlayServer
+	logger *zap.Logger
+}
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: formatting, i.e. enable format on save

File: pkg/datarepair/checker/checker.go [hunk 2]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+		overlay: overlay,
+		logger: logger,
+	}
+}
+
+// IdentifyInjuredSegments checks for missing pieces off of the pointerdb and overlay cache
+func (c *Checker) IdentifyInjuredSegments(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	c.logger.Debug("entering pointerdb iterate")
+	err = c.pointerdb.Iterate(storage.IterateOptions{Prefix: storage.Key(c.params.Prefix), First: storage.Key(c.params.First), Recurse: c.params.Recurse, Reverse: c.params.Reverse},
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: Split the options on different lines.

```
storage.IterateOptions{
	Prefix: storage.Key(c.params.Prefix),
	First: storage.Key(c.params.First),
	Recurse: c.params.Recurse,
	Reverse: c.params.Reverse,
}, func(it storage.Iterator) {
```

File: pkg/pb/datarepair.proto [hunk 3]
```diff
@@ -10,4 +10,14 @@ package repair;
...
     repeated int32 lost_pieces = 2;
+    repeated int32 healthy_pieces = 3;
 }
+
+//IdentifyRequest is a request message for the identifyInjuredSegments private method
+message IdentifyRequest {
+    string prefix = 1;
+    string first = 2;
+    bool recurse = 3;
+    bool reverse = 4;
```
Comment (egonelbre) [inline pkg/pb/datarepair.proto]: Are all these really required for `Identify`?

File: pkg/datarepair/checker/checker.go [hunk 4]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+				err = proto.Unmarshal(item.Value, pointer)
+				if err != nil {
+					return checkerError.New("error unmarshalling pointer %s", err)
+				}
+				pieces := pointer.Remote.RemotePieces
+				var nodeIDs []dht.NodeID
+				for _, p := range pieces {
+					nodeIDs = append(nodeIDs, kademlia.StringToNodeID(p.NodeId))
+				}
+				missingPieces, healthyPieces, err := c.offlineAndOnlineNodes(ctx, nodeIDs)
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: `offlinePiece, onlinePieces`... We don't know if they are healthy until we try to download them. Data can get corrupt.

File: pkg/datarepair/checker/checker.go [hunk 5]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+				if err != nil {
+					return checkerError.New("error getting missing offline nodes %s", err)
+				}
+				if int32(len(missingPieces)) >= pointer.Remote.Redundancy.RepairThreshold {
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: RepairThreshold is in terms of available `pieces` not `missing`. So, `int32(len(offlinePieces)) <= pointer.Remote.Redundancy.RepairThreshold`.

File: pkg/datarepair/checker/checker.go [hunk 6]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+					err = c.repairQueue.Enqueue(&pb.InjuredSegment{
+						Path: string(item.Key), 
+						LostPieces: missingPieces,
+						HealthyPieces: healthyPieces,
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: Does the repairer need this information? It could try downloading from all of the nodes instead, because when it's repairing the status of the pieces might have changed.

File: pkg/datarepair/checker/checker.go [hunk 7]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+				}
+			}
+			return nil
+		},
+	)
+	return err
+}
+
+// returns the indices of offline and online nodes
+func (c *Checker) offlineAndOnlineNodes(ctx context.Context, nodeIDs []dht.NodeID) (offline []int32, online []int32, err error) {
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: Returning count of online pieces would probably suffice.

File: pkg/datarepair/checker/checker.go [hunk 8]
```diff
@@ -19,3 +30,107 @@ func (c *Config) Run(ctx context.Context) (err error) {
 
 	return err
 }
+
+
+// Checker contains the information needed to do checks for missing pieces
+type Checker struct {
+	params *pb.IdentifyRequest
+	pointerdb storage.KeyValueStore
```
Comment (egonelbre) [inline pkg/datarepair/checker/checker.go]: I'm not sure we want direct access to the underlying database of pointerdb. This will cause hidden coupling in terms of how data is stored in the database. The requests should go against the pointerdb API.

Comment (egonelbre) [reply pkg/datarepair/checker/checker.go]: Although that might need to happen eventually, I would prefer we do it against the API for now.

File: pkg/pb/datarepair.proto (hunk 3 above)
Comment (jenlij) [reply pkg/pb/datarepair.proto]: not necessarily, i wasnt sure what we would need eventually.

File: pkg/datarepair/checker/checker.go (hunk 4 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: That is a good point. However, there was discussion with @jtolds  and @kaloyan-raev that we could take that risk?

File: pkg/datarepair/checker/checker.go (hunk 5 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: Ok that makes sense, that we always need # healthy pieces greater than the repair threshold

File: pkg/datarepair/checker/checker.go (hunk 6 above)
Comment (egonelbre) [reply pkg/datarepair/checker/checker.go]: sounds good

File: pkg/datarepair/checker/checker.go (hunk 4 above)
Comment (egonelbre) [reply pkg/datarepair/checker/checker.go]: I just mean the naming... yes, we kind of consider them "healthy" unless proven otherwise. Just trying to keep consistency with the method name.

File: pkg/datarepair/checker/checker.go (hunk 5 above)
Comment (kaloyan-raev) [reply pkg/datarepair/checker/checker.go]: Wait... I am confused. We should be adding a segment to the repair queue if the number of healthy pieces are below the repair threshold, i.e.
```go
int32(len(healthyPieces)) < pointer.Remote.Redundancy.RepairThreshold
```

File: pkg/datarepair/checker/checker.go [hunk 9]
```diff
@@ -19,3 +30,110 @@ func (c *Config) Run(ctx context.Context) (err error) {
...
+				pieces := pointer.Remote.RemotePieces
+				var nodeIDs []dht.NodeID
+				for _, p := range pieces {
+					nodeIDs = append(nodeIDs, kademlia.StringToNodeID(p.NodeId))
+				}
+				missingPieces, healthyPieces, err := c.offlineAndOnlineNodes(ctx, nodeIDs)
+				if err != nil {
+					return checkerError.New("error getting missing offline nodes %s", err)
+				}
+				if int32(len(missingPieces)) <= pointer.Remote.Redundancy.RepairThreshold {
```
Comment (kaloyan-raev) [inline pkg/datarepair/checker/checker.go]: Not sure why we are comparing the number of missing pieces with the repair threshold. We should be adding a segment to the repair queue if the number of healthy pieces are below the repair threshold, i.e.
```go
int32(len(healthyPieces)) < pointer.Remote.Redundancy.RepairThreshold
```

File: pkg/datarepair/checker/checker.go (hunk 5 above)
Comment (egonelbre) [reply pkg/datarepair/checker/checker.go]: Yes, sorry, my mistake in the code part.

File: pkg/datarepair/checker/checker.go (hunk 9 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: Is this what the white paper says? I had it as len(missingpieces) >= repairthreshold before.

File: pkg/datarepair/checker/checker.go [hunk 10]
```diff
@@ -7,17 +7,23 @@ import (
...
+	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/provider"
+	"storj.io/storj/storage"
 )
 
 var (
 	mon = monkit.Package()
 	// Error is a standard error class for this package.
-	Error = errs.Class("checker error")
+	checkerError = errs.Class("checker error")
```
[the remark this answers is not in the document]
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: sure that works

File: pkg/datarepair/checker/checker.go [hunk 11]
```diff
@@ -28,7 +27,7 @@ type Config struct {
 
 // Run runs the checker with configured values
 func (c Config) Run(ctx context.Context, server *provider.Provider) (err error) {
-	defer mon.Task()(&ctx)(&err)
+	defer datarepair.Mon.Task()(&ctx)(&err)
```
Comment (kaloyan-raev) [inline pkg/datarepair/checker/checker.go]: This looks weird. Why did you move the `var mon = monkit.Package()` declaration to datarepair/common.go? Why not under the same package, i.e. checker/common.go?

File: pkg/datarepair/checker/checker.go [hunk 12]
```diff
@@ -51,3 +50,105 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
 
 	return server.Run(ctx)
 }
+
+// Checker contains the information needed to do checks for missing pieces
+type Checker struct {
+	params      *pb.IdentifyRequest
+	pointerdb   storage.KeyValueStore
+	repairQueue *queue.Queue
+	overlay     pb.OverlayServer
```
Comment (kaloyan-raev) [inline pkg/datarepair/checker/checker.go]: Shouldn't we actually use the `overlay.Client` interface instead of the "server" one? The repair checker is a client to the overlay cache, isn't it?

File: pkg/pb/datarepair.proto [hunk 13]
```diff
@@ -10,4 +10,14 @@ package repair;
 message InjuredSegment {
     string path = 1;
     repeated int32 lost_pieces = 2;
+    repeated int32 healthy_pieces = 3;
```
Comment (kaloyan-raev) [inline pkg/pb/datarepair.proto]: Still puzzled why we need both lost and healthy pieces in the injured segment.

File: pkg/datarepair/checker/checker.go (hunk 11 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: I can move it to checker/common. Thought to move it to the data repair checker since all of the packages within that use mon.task, but I do agree it is clunky

File: pkg/datarepair/checker/checker.go (hunk 12 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: JT said since the checker will run on the server, the overlay server works here. It also allows us to test with the mock cache.
